cni/pkg/ebpf/app/ambient_redirect.bpf.c

```diff
@@ -0,0 +1,277 @@
+            return TC_ACT_OK;
+        udph = data + sizeof(*eth) + sizeof(*iph);
+        if (udph->dest == bpf_htons(UDP_P_DNS))
+            return TC_ACT_OK;
+    }
+
+    __builtin_memcpy(eth->h_dest, zi->mac_addr, ETH_ALEN);
+
+    skb->cb[4] = OUTBOUND_CB;
+
+    dbg("outbound redirect to ztunnel ifidx: %u\n", zi->ifindex);
+    return bpf_redirect(zi->ifindex, 0);
+}
+
+// For app pod veth pair(in host ns) egress hook
+SEC("tc")
+int app_inbound(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct ethhdr  *eth = data;
+    struct iphdr *iph = NULL;
+    struct udphdr *udph;
+    struct ztunnel_info *zi = NULL;
+
+    if (skb->cb[4] == BYPASS_CB)
+        return TC_ACT_OK;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support ipv6
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP && iph->protocol != IPPROTO_UDP)
+        return TC_ACT_OK;
+
+    zi = get_ztunnel_info();
+
+    if (!zi || zi->ifindex == 0) {
+        dbg("inbound_to_ztunnel unable to retrieve ztunnel_info\n");
+        return TC_ACT_OK;
+    }
+
+    if (iph->protocol == IPPROTO_UDP) {
+        if (data + sizeof(*eth) + sizeof(*iph) + sizeof(*udph) > data_end)
+            return TC_ACT_OK;
+        udph = data + sizeof(*eth) + sizeof(*iph);
+        if (udph->source == bpf_htons(UDP_P_DNS))
+            return TC_ACT_OK;
+    }
+
+    __builtin_memcpy(eth->h_dest, zi->mac_addr, ETH_ALEN);
+
+    skb->cb[4] = INBOUND_CB;
+
+    dbg("inbound redirect to ztunnel ifidx: %u\n", zi->ifindex);
+    return bpf_redirect(zi->ifindex, 0);
+}
+
+// For ztunnel pod veth pair(in host ns) ingress hook
+SEC("tc")
+int ztunnel_host_ingress(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    struct ethhdr  *eth = data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct iphdr *iph = NULL;
+    struct app_info *pi = NULL;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support ipv6
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP && iph->protocol != IPPROTO_UDP)
+        return TC_ACT_OK;
+
+    pi = get_app_info_from_ipv4(iph->daddr);
+    if (!pi) {
+        // ip is not in ambient managed mesh
+        // dbg("ztunnel_host_ingress unable to retrieve pod info: 0x%x\n", bpf_ntohl(iph->daddr));
+        return TC_ACT_OK;
+    }
+
+    __builtin_memcpy(eth->h_dest, pi->mac_addr, ETH_ALEN);
+    skb->cb[4] = BYPASS_CB;
+
+    dbg("ztunnel redirect to app(0x%x) ifidx: %u\n", bpf_ntohl(iph->daddr), pi->ifindex);
```

Comment:
Why not use `bpf_redirect/bpf_redirect_to_peer` here, as they only require `ifindex`?

Reply:
bpf_redirect_peer requires newer kennel version

Reply:
It's in 5.x.x -> was there a reason we picked `4.20` as the lower bound in the original proposal?

Reply:
This again I think is something that needs followup (but doesn't need to block this specific PR) - eBPF progs practically speaking require us to declare the kernel versions we support. 

-We should test (at least) Istio eBPF CNI releases with the upper and lower bound kernels we claim to support. 

-Our upper and lower bounds should be LTS releases that are actively supported at time of Istio release - 4.20 is neither (was EOL'd in 2019) which means the lower bound we probably should actually document/claim/support is kernel 5.4.

- Newer kernels support faster/better BPF APIs - we have to decide under what conditions we make use of those (alt progs per-kernel-version? Or we don't use any APIs not present in our lower bound?)

Reply:
Naive question, can you do something like
```
if kernel == old { slow } else { fast } 
```

in the bpf code?

Reply:
And same question on 4.20 - does 4.19 (LTS) work?

Reply:
@howardjohn 

1. No. You can do feature probes with e.g. `bpftool` or from the userspace prog loader before loading, but IIRC given the way the verifier works you cannot really dynamically pick different runttime logic in the kernelspace prog.

2. Unsure - if 4.19 works that's fine - but do we want to support that low of a bound? It does constrain what the progs can do, and a very wide bound (could) make testing the ends of both bounds a bit harder.

Reply:
Not necesarily, but if we are going to pick 4.20 we should pick 4.19 instead was my main point. I am doing some research into the minimum version our customers use at least - would be interested to hear from others as well.

Reply:
That makes sense - as long as it's something we pick with intent. My $0.02/input is that 5.4 is probably a pretty good lower bound - it's 2 years old, most of the major distros are using at least that in most of their extant LTS releases.

Reply:
```
but IIRC given the way the verifier works you cannot really dynamically pick different runtime logic in the kernelspace prog.
```

retroactively - old me is completely wrong, and you can do this: https://nakryiko.com/posts/bpf-core-reference-guide/#handling-incompatible-field-and-type-changes